Add cmake setup for building

SteveBronder opened the pull request:
This adds a full cmake setup for building. I added some docs to the `readme.md` in `walnuts_cpp` to make it clear how to use cmake. CMake allows for some niceties.

As an example, here is how you build the project with cmake and run the nuts test and the little benchmark I added for nuts.

```bash
# Call cmake with our source as our current directory and build in a new folder "build"
cmake -S . -B "build" -DCMAKE_BUILD_TYPE=RELEASE 
cd build
# List possible targets
make help
make test_nuts
./test_nuts
```

The cmake setup is nice because it also handles dependencies for us like Eigen. For testing and benchmarking, cmake will optionally setup google benchmark and google test for us as well.

- Moved `nuts.hpp` to `include/nuts`
- Moved `tests.cpp` to `examples` folder
- Adds a `benchmark` library so we can benchmark with google bench
- Adds an example for running `perf record` and `perf report` from cmake
- Adds `FlameGraph` and `perf` cmake commands for profiling executables

SteveBronder commented:
@WardBrian could you add the setup here so that we can build `test_stan.cpp`?

WardBrian commented:
Building test_stan itself should be super easy, it just needs the one header. Building a model to actually use with test_stan is more of an ordeal. Do you want that? If so, might make sense to merge this without and look into it separately, it would probably be easiest to make a branch of BridgeStan using your math branch that has a cmakelists.txt already

SteveBronder commented:
> Building test_stan itself should be super easy, it just needs the one header. Building a model to actually use with test_stan is more of an ordeal. Do you want that?

Let's do all that in another PR

bob-carpenter commented:
I have zero opinions on builds and will leave this up to you.  But when you think it's ready to merge, I would very much like to use myself as a guinea pig to see if I can follow the instructions.  Right now, they seem to be missing where to get `benchmarks.h` for example.  

Thanks for writing doc.  I'd feel better using it if I knew what all those command line arguments were doing in things like

```
cmake -S . -B "build" -DCMAKE_BUILD_TYPE=RELEASE -DWALNUTS_BUILD_TESTS=ON -DWALNUTS_BUILD_BENCHMARKS=ON
```

SteveBronder commented:
Thanks Bob! For the command line arguments I recommend looking at `man cmake`. I'll make the comment on the line above the one you posted a bit more verbose. I've updated it in the below and will update it in the readme.

Could you send me any errors you get from running the following? I think I need to get the CI for PRs on here so that we can make sure all this stuff works across different operating systems.

```bash
# Call cmake with our source '-S .' as our current directory and build '-B "build"'in a new folder "build"
cmake -S . -B "build" -DCMAKE_BUILD_TYPE=RELEASE -DWALNUTS_BUILD_TESTS=ON -DWALNUTS_BUILD_BENCHMARKS=ON
cd build
# List possible targets
make help
make VERBOSE=1 -j3 normal_nuts
# After building
./benchmarks/normal_nuts
make VERBOSE=1 -j3 test_nuts
./tests/test_nuts
```

I also added a perf directory that makes it pretty easy to generate nice flame graphs from `perf record` output. See the link [here](https://gist.githack.com/SteveBronder/a1528a0a47ae136e926ecca8255a7fba/raw/5f1b3fef7d1fb0bc92be57e25da4556ec1e03104/perf_test_nuts.svg) for a pretty neat flame graph!

bob-carpenter commented:
Two things.

1.  I don't understand what the flame graph's supposed to be plotting.

2.  The more CI and build details pile on, the more frustrated I'm going to get if anything ever stops working for me.  I am terrible at debugging config and don't understand either cmake or Jenkins configs.  This is why I eventually stopped coding in Stan---the infrastructure just overwhelmed me.

3.  (Bonus thing). Can we just include all the libraries we need in the GitHub?  It looks like you're bringing in more dependencies and I want people with my level of build expertise or less to be able to get this off the ground without too much head scratching about which versions of packages to download and where to put them.

bob-carpenter commented:
What does "vendor Eigen" mean?  I forgot about BridgeStan being a dependency---I haven't caught up to that yet.

WardBrian commented:
Vendoring is the process of including your dependencies in your own source distribution rather than relying on something like a package manager.

Bridgestan should probably stay an optional dependency forever, since the algorithm itself won’t ever need it, so it’s in a difference class than Eigen I think

SteveBronder commented:
> I don't understand what the flame graph's supposed to be plotting.

In short, they take the perf recording and show how much time the process spends in each function and subfunctions. So if you hover over `nuts::nuts` You'll see the program spend about 76% of the time in the `nuts::nuts` function. Clicking on `nuts::nuts`'s bar will slice into a subgraph so you can zoom into that call site to see what the program spent the most time on within that function.

You can read more about them [here](https://www.brendangregg.com/flamegraphs.html)

> The more CI and build details pile on, the more frustrated I'm going to get if anything ever stops working for me. I am terrible at debugging config and don't understand either cmake or Jenkins configs. This is why I eventually stopped coding in Stan---the infrastructure just overwhelmed me.

imo if we don't use cmake we will end up in the same place our stan make files are. CMake has its annoyances, but it is way easier to debug than raw make.

> (Bob) (Bonus thing). Can we just include all the libraries we need in the GitHub? It looks like you're bringing in more dependencies and I want people with my level of build expertise or less to be able to get this off the ground without too much head scratching about which versions of packages to download and where to put them.

> (Brian) I think additional dependencies for things like benchmarking are fine, as most devs can avoid them. It should also be trivial to keep a normal makefile around for basic things I would expect.

Besides `Eigen` all the dependencies are behind flags. So if someone does not want to do benchmarking they never have to install google bench. Same for the tests. If someone does not want to build our tests then they set `WALNUTS_BUILD_TESTS=OFF` and cmake never even downloads google test. All the extra dependencies are behind flags so if someone does not use a feature it's dependencies will not be downloaded.

> (Brian) We can and probably should vendor Eigen, I’d be a bit more apprehensive about keeping a copy of BridgeStan in the repository
 
Do you mean having a hard copy of Eigen in the repo. If so why is that your preference? imo I prefer using cmake's dependency management. I use it for several other projects and it's made life a lot easier.

WardBrian commented:
For header only libraries I think it’s easier if the git clone and (eventual) release tarballs just contain all the code you need. Especially when people are working behind firewalls I think it can cause them problems when the build step wants to make web requests 

Besides that, it’s mostly for personal preference reasons; if I want to browse through the code on e.g. github it’s nice if I can also see the Eigen code

SteveBronder commented:
@WardBrian actually cmake has a package called cpack that handles the case where you want to package your third party sources with your code for releases. I just added that to the cmake. It builds a tarball with `Eigen` and `walnuts`

bob-carpenter commented:
I'm happy to follow any process if it works.

WardBrian commented:
cpack looks neat, that’s new to me.

At any rate, I definitely think CMake is the right way to go, both for cross platform support and because some options of python integration like Pybind play very nicely with it. There are just some choices one can make while using CMake that determine how easy or hard it is to use *anything else* as a secondary option, but I think we can reassess those over time

bob-carpenter commented:
Is this all necessary?  There's now more make code than C++ code.

Also, is there a way to do this without putting the code under `include`?

bob-carpenter commented:
The top-level README needs to be redone to include the new C++ stuff.

SteveBronder commented:
> I am really bad at deeply nested directory structure. I have trouble seeing the trees through a forest this dense. The boilerplate feels like it's taking over here.

This looks like boilerplate, but it is just automating all the stuff I'd normally do manually. For example, say I want to run perf and get which call sites the program is sitting in for the longest. Normally I'd have to run 

```bash
clang++ -std=c++17 -O3 -I lib/eigen-3.4.0 test.cpp -o test
perf record -F 99 -g --call-graph dwarf  -o ./data/perf_test.data -- test
perf report -g --call-graph graph --stdio --stdio-color=always -i ./data/perf_test.data
```

Then if I want to have a plot of that perf data I then need to run 
```bash
perf script -i  ./data/perf_test.data | \
  ../FlameGraph/stackcollapse-perf.pl | \
  ../FlameGraph/flamegraph.pl --title='Nuts Graph' > ./data/nuts.svg
```

but with all this I can just call

```bash
# /usr/bin/bash
# in walnuts_cpp directory
cmake -S . -B "build"
# Prints the report showing the callsites
cmake --build build target perf_report_test_nuts
# makes the svg showing the time spent at each callsite
cmake --build build target flamegraph_test_nuts
```

And now everyone else who wants to call perf can just use the cmake calls above.

> Use $ or something to indicate that what you're doing is shell commands. And say as much beforehand.

I will indicate they are bash with a comment, but I like people to be able to copy and paste so I'm leaving off the $ at the beginning.

> Give an example of which directory you start in before issuing the first command. That's just cd walnuts_cpp here, but I hate having to guess this up front.

I think that's a good idea and added that to the top of each cmake example

> All of these commands should at least give a hint of their results and what they do and how long they're going to take to run.

I added a bunch of info to say what the results of the operations are.

> I like space between commands so I don't get lost, but that's optional. I'd actually have each of these separated by text that says what it does, but that's because I always have to count on my fingers to figure these things out.

I have a lot of comments between things now but I'm fine with space. I'm not sure where you would want space in particular so feel free to make any changes you would like.

> Mention that the build command will go out on the internet and build dependencies, so it may take a minute and will generate a ton of output, presumably none of which matters? I never know what to do with programs giving me a ton of output I don't know how to interpret. So happy this one worked first try---this is where they always fail on me.

It matters if something goes wrong, but mostly just tells you what options cmake is making available for the dependency. So for a standard user that output does not matter.

bob-carpenter commented:
> now everyone else who wants to call perf can just use the cmake calls above.

I know, but on the dev side we now have to doc it and maintain it going forward.  On the user side, it's just one more piece of complexity in the scripts/environment.

I'm not saying don't do it, but please consider the cost.

> I like people to be able to copy and paste so I'm leaving off the $ at the beginning.

Good point.   Then I guess I'll take back the request to show output as that would get super confusing.

bob-carpenter commented:
> And now everyone else who wants to call perf can just use the cmake calls above.

One issue I've always had with frameworks like this is that they obfuscate as well as automate specific calls.  When everyone wants to do the exact same thing, like build a piece of software, it's great.  But if not, I find it gets in the way.  It can wind up being a lot harder to modify a call with another system in the middle because now you have to understand the thing being called and the middleware.

At some point, there will be a crossover because the calls get so complicated that you couldn't formulate them manually.   I'm just griping because I don't think we were there yet.  We had one .hpp and two .cpp files and a single dependency on Eigen and now it's this big sprawling directory of config files.

Partly I'm just griping because in this case the middleware is `cmake`, which I don't know and have historically struggled with because I'm really really bad at dev ops.  It feels like a lot of homework to understand how to build a couple files.  I'm sure it'll be good for me long run, but it's already slowing me down on this project.

bob-carpenter commented:
Sounds good.  Thanks for trimming.
